testsuite/tests/typing-misc/constraints.ml

```diff
@@ -433,3 +433,16 @@ and r = [s cstr t | `Bar]
 type s = int
 and r = [ `Bar | `Foo ]
 |}]
+
+
+(* PR #12334 *)
+
+type 'a t = 'a foo foo
+and 'a foo = int constraint 'a = int
+[%%expect{|
+Line 1, characters 12-22:
+1 | type 'a t = 'a foo foo
+                ^^^^^^^^^^
+Error: Constraints are not satisfied in this type.
+       Type "'a foo foo" should be an instance of "int foo"
```

Comment:
It would be very nice to get an additional

```
Constraints cannot be inferred within a mutually recursive group of definitions
```

or something. Because `'a foo foo` *is* actually an instance of `int foo`.

Not sure how hard it would be to arrange for the better message.

Reply:
Having an exact error message is going to be difficult here. We could track the recursive use, and point it as a potential cause.

Reply:
I think that tracking the recursive use would be very helpful, if it's not too hard.